doobie/src/main/scala/tamer/db/Setup.scala

```diff
@@ -0,0 +1,84 @@
+package tamer.db
+
+import com.sksamuel.avro4s.{Decoder, Encoder, SchemaFor}
+import doobie.Query0
+import tamer.Serde
+import zio.UIO
+
+import java.time.{Duration, Instant}
+
+trait QueryBuilder[V, -S] {
+
+  /** Used for hashing purposes
+    */
+  val queryId: Int
+  def query(state: S): Query0[V]
+}
+
+trait HashableState {
+
+  /**  It is required for this hash to be consistent even across executions
+    *  for the same semantic state. This is in contrast with the built-in
+    *  `hashCode` method.
+    */
+  val stateHash: Int
+}
+
+final case class ResultMetadata(queryExecutionTime: Long)
+final case class QueryResult[V](metadata: ResultMetadata, results: List[V])
+
+case class Setup[
+    K <: Product: Encoder: Decoder: SchemaFor,
+    V <: Product: Encoder: Decoder: SchemaFor,
+    S <: Product with HashableState: Encoder: Decoder: SchemaFor
+](
+    queryBuilder: QueryBuilder[V, S],
+    override val defaultState: S,
+    keyExtract: V => K,
+    stateFoldM: S => QueryResult[V] => UIO[S]
+) extends tamer.Setup[K, V, S](
```

Comment:
this _may_ be required
```suggestion
) extends _root_.tamer.Setup[K, V, S](
```